beacon_node/http_api/src/state_id.rs

```diff
@@ -94,35 +112,46 @@ impl StateId {
                 warp_utils::reject::custom_not_found(format!("beacon state at slot {}", slot))
             })?;
 
-        Ok((root, execution_optimistic))
+        Ok((root, execution_optimistic, finalized))
     }
 
     /// Return the `fork` field of the state identified by `self`.
     /// Also returns the `execution_optimistic` value of the state.
     pub fn fork_and_execution_optimistic<T: BeaconChainTypes>(
         &self,
         chain: &BeaconChain<T>,
-    ) -> Result<(Fork, bool), warp::Rejection> {
-        self.map_state_and_execution_optimistic(chain, |state, execution_optimistic| {
-            Ok((state.fork(), execution_optimistic))
+    ) -> Result<(Fork, bool, bool), warp::Rejection> {
+        self.map_state_and_execution_optimistic(chain, |state, execution_optimistic, finalized| {
+            Ok((state.fork(), execution_optimistic, finalized))
         })
     }
 
+    /// Return the `fork` field of the state identified by `self`.
+    /// Also returns the `execution_optimistic` value of the state.
+    /// Also returns the `finalized` value of the state.
+    pub fn fork_and_execution_optimistic_and_finalized<T: BeaconChainTypes>(
+        &self,
+        chain: &BeaconChain<T>,
+    ) -> Result<(Fork, bool, bool), warp::Rejection> {
+        let (state, execution_optimistic, finalized) = self.state(chain)?;
```

Comment:
I think this might be it, lets use `map_state_and_execution_optimistic` instead of `self.state` and see if that improves the `head_state_fork` benchmark.

The reason it's probably better to use the mapping function is that it avoids cloning any beacon states. The head beacon state is at least a few hundred MB in memory and takes on the order of 20-50ms to clone depending on the machine. By using the `map` function we can just pull out the `state.fork()` without cloning the state.